api_design/DDS-974-enforce_upload_chunk_size.md

```diff
@@ -0,0 +1,89 @@
+# DDS-974 Enforce Upload Chunk Size
+
+## Deployment View
+
+N/A
+
+## Logical View
+
+A critical issue was recently exposed with the chunked upload capability.  The  storage provider (Swift), is configured with a set maximum number of segments (i.e. Data Service chunks), beyond which it cannot coalesce a large file.  To remedy this situation, the API will be extended to enforce a minumum chunk size, based on the overall file size and max segments setting.
+
+#### Summary of impacted APIs
+
+|Endpoint |Description |
+|---|---|
+| `GET /storage_providers` | List supported storage providers. |
+| `GET /storage_providers/{id}` | Get a storage provider. |
+| `POST /projects/{id}/uploads` | Inititate a chunked upload. |
+| `PUT /uploads/{id}/chunks` | Generate and return a pre-signed upload URL for a chunk.  |
+
+#### API Specification
+This section defines the proposed API interface extensions.
+
+##### List supported storage providers / Get a storage provider
+`GET /storage_providers` / `GET /storage_providers/{id}`
+
+###### Response Example (Extensions)
+The following properties will be added to the storage providers resource:
+
++ **chunk\_max\_size\_bytes** - The maximum size of a chunk that can be upload.
++ **chunk\_max\_number** - The maximum number of chunks that can be uploaded for a single file. 
++ **file\_max\_size\_bytes** - Maximum supported file size that can be uploaded. (`chunk_max_size_bytes * chunk_max_number`)
+
+```
+{
+      "id": "g5579f73-0558-4f96-afc7-9d251e65bv33",
+      "name": "duke_oit_swift",
+      "description": "Duke OIT Storage",
+      "chunk_hash_algorithm": "md5",
+      "chunk_max_size_bytes": 5368709120,
+      "chunk_max_number": 1000,
+      "file_max_size_bytes": 5497558138880,    
+      "is_deprecated": false
+}
+```
+
+##### Intitate a chunked upload
+`POST /projects/{id}/uploads`
+
+###### Response Headers (Extensions)
+The following custom response headers will be added to inform clients of the minimum chunk size that may be utlized to ensure chunks can be coalesced, as well as the maximum chunk size the storage provider can accommodate.
+
++ **x-min-chunk-upload-size** - The minimum chunk size in bytes.
++ **x-max-chunk-upload-size** - The maximum chunk size in bytes.
+
+##### Generate and return a pre-signed upload URL for a chunk
+`PUT /uploads/{id}/chunks`
+
+###### Response Messages (Extensions)
```

Comment:
We'll need one more 400 response for when `PUT /uploads/{id}/chunks` tries to create a chunk on an upload that has `chunk_max_number` of chunks.